### PR TITLE
internal: add ban manager.

### PR DESCRIPTION
**Pull request by dnldd**
This refactors peer banning functionality and ban score tracking into a ban manager. Associated tests have been added.

Integrating the ban manger will be done in a subsequent PR.

### CODE REVIEW

**Review by davecgh — CHANGES_REQUESTED** [review]
Aside from the fact this is breaking the API of `connmgr` without being associated with a major module version bump, I don't believe the ban score can or should be moved anyway.  It is specifically related to connection managed and it is something that wallet also needs and uses.

File: peer/peer.go [hunk 1]
```diff
@@ -433,6 +433,11 @@ type Peer struct {
 	cfg     Config
 	inbound bool
 
+	// These fields track the peer's whitelisted status and ban score.
+	isWhitelisted bool
+	banScore      *DynamicBanScore
+	banManager    BanManager
```
Comment (davecgh) [inline peer/peer.go:439]: Similarly, this doesn't make sense to me. The ban manager, as it name implies, is the thing that manages peers, not the other way around.

File: server.go [hunk 2]
```diff
@@ -3655,6 +3620,15 @@ func newServer(ctx context.Context, listenAddrs []string, db database.DB, chainP
 	}
 	s.connManager = cmgr
 
+	bmgr := connmgr.NewBanManager(&connmgr.BanManagerConfig{
```
Comment (davecgh) [inline server.go:3623]: Can just do `s.banManager = ` here.

File: peer/peer.go [hunk 3]
```diff
@@ -433,6 +433,11 @@ type Peer struct {
 	cfg     Config
 	inbound bool
 
+	// These fields track the peer's whitelisted status and ban score.
+	isWhitelisted bool
```
Comment (davecgh) [inline peer/peer.go:437]: This doesn't make sense to me to be on the peer.  It is a function of the ban manager and is something that should be housed by it.

File: peer/peer.go [hunk 4]
```diff
@@ -2076,6 +2091,36 @@ func (p *Peer) AssociateConnection(conn net.Conn) {
 	}(p)
 }
 
+// SetIsWhitelisted updates the whitelisted status of the peer based on
+// the provided peer whitelist.
+//
+// This should only be called after AssociateConnection ensuring the peer
+// references its underlying connection.
+func (p *Peer) SetIsWhitelisted(whitelist []*net.IPNet) {
```
Comment (davecgh) [inline peer/peer.go:2099]: Per above, I don't understand why this would be here.  The ban manager is the thing that needs to control whitelisting.

**Review by davecgh — COMMENTED** [review]
Thanks for the PR.  While I definitely understand and agree with the notion of a ban manager, perhaps I'm just being a bit dense, but, as I further describe inline, I really don't understand the overall design here.  It seems to me like everything is exactly backwards.  I would expect the ban manager to manage the banning of the peers, but instead it seems like there is a whole bunch of things that I would expect to be the responsibility of the ban manager, like whitelisting and tracking of the ban score, being handled by each individual peer.

File: server.go [hunk 5]
```diff
@@ -557,6 +558,7 @@ type serverPeer struct {
 func newServerPeer(s *server, isPersistent bool) *serverPeer {
 	return &serverPeer{
 		server:         s,
+		banManager:     s.banManager,
```
Comment (davecgh) [inline server.go:561]: I would not put this on every peer.  It takes up additional space for each one and isn't really needed.  Instead, I'd suggest just updating `addBanScore` to be something like:

```Go
func (sp *serverPeer) addBanScore(persistent, transient uint32, reason string) bool {
	return sp.s.banManager.AddBanScore(sp.Peer, persistent, transient, reason)
}
```

That way the rest of the code can even stay the same.

File: server.go [hunk 6]
```diff
@@ -1102,7 +1067,7 @@ func (sp *serverPeer) OnBlock(p *peer.Peer, msg *wire.MsgBlock, buf []byte) {
 func (sp *serverPeer) OnInv(p *peer.Peer, msg *wire.MsgInv) {
 	// Ban peers sending empty inventory requests.
 	if len(msg.InvList) == 0 {
-		sp.server.BanPeer(sp)
+		sp.banManager.BanPeer(p)
```
Comment (davecgh) [inline server.go:1070]: `sp.s.banManager.BanPeer(p)` if suggested change is made.  Or you could even introduce a helper like:

```Go
func (sp *serverPeer) banPeer() {
	sp.s.banManager.BanPeer(sp.Peer)
}
```

Then all these ban calls would be just turn into `sp.banPeer()`

File: server.go [hunk 7]
```diff
@@ -1481,14 +1454,14 @@ func (sp *serverPeer) OnNotFound(p *peer.Peer, msg *wire.MsgNotFound) {
...
-		if sp.addBanScore(20*numBlocks, 0, reason) {
+		if sp.banManager.AddBanScore(p, 20*numBlocks, 0, reason) {
 			return
 		}
 	}
 	if numTxns > 0 {
 		txStr := pickNoun(uint64(numTxns), "transaction", "transactions")
-		reason := fmt.Sprintf("%d %v not found", numTxns, txStr)
-		if sp.addBanScore(0, 10*numTxns, reason) {
+		reason := fmt.Sprintf("%d %v not found", numBlocks, txStr)
```
Comment (davecgh) [inline server.go:1463]: Looks like a rebase reversion.

**Review by davecgh — CHANGES_REQUESTED** [review]
This looks much better with the updates.  I have various comments for some outstanding things and once those are addressed I think this will be good enough.  There are various other improvements I can see that would be nice for the future, but I think it's better to just focus on getting the broad strokes in first.

File: server.go [hunk 8]
```diff
@@ -657,45 +656,18 @@ func (sp *serverPeer) pushAddrMsg(addresses []*addrmgr.NetAddress) {
 	sp.addKnownAddresses(knownNetAddrs)
 }
 
+// banPeer bans the provided peer.
```
Comment (davecgh) [inline server.go:659]: There is no provided peer, rather it is the peer itself.

That said, I might suggest something a little more descriptive such as:

```suggestion
// banPeer notifies the ban manager associated with the server to ban the peer.
```

File: connmgr/banmanager.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+	*peer.Peer
+
+	isWhitelisted bool
+}
+
+// BanManager represents a peer ban score tracking manager.
+type BanManager struct {
+	cfg       BanManagerConfig
+	peers     map[string]*banMgrPeer
+	banScores map[string]*DynamicBanScore
```
Comment (davecgh) [inline connmgr/banmanager.go:48]: I would expect this to live in `banMgrPeer` as `banScore DynamicBanScore` (notice it's also not a pointer in the struct which results in less heap objects) since the purpose of that struct is to house per-peer state which is what this is.

File: connmgr/banmanager.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+	return &BanManager{
+		cfg:       *cfg,
+		peers:     make(map[string]*banMgrPeer, cfg.MaxPeers),
+		banScores: make(map[string]*DynamicBanScore, cfg.MaxPeers),
+	}
+}
+
+// IsPeerWhitelisted checks if the provided peer is whitelisted.
+func (bm *BanManager) IsPeerWhitelisted(p *peer.Peer) bool {
+	host, _, err := net.SplitHostPort(p.Addr())
```
Comment (davecgh) [inline connmgr/banmanager.go:63]: This doesn't appear to be doing what we discussed.  I would expect this exported version to lookup the peer in the `peers` map (under the mutex of course) and return the field.

e.g. Assuming the other suggested changes:

```Go
func (bm *BanManager) IsPeerWhitelisted(p *peer.Peer) bool {
	bm.mtx.Lock()
	bmp := bm.lookupPeer(p)
	bm.mtx.Unlock()
	if bmp == nil {
		return false
	}
	return bmp.isWhitelisted
}
```

This code is obviously needed in order to originally set the field, but I would expect it to be unexported.

File: connmgr/banmanager.go [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+		return false
+	}
+	bm.mtx.Unlock()
+
+	if bmp.isWhitelisted {
+		log.Debugf("Misbehaving whitelisted peer %s: %s", p, reason)
+		return false
+	}
+
+	banScore, ok := bm.banScores[p.Addr()]
```
Comment (davecgh) [inline connmgr/banmanager.go:171]: Naturally this all would just become `banScore := &bmp.banScore` with the changes mentioned elsewhere.

File: connmgr/banmanager.go [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+		score := banScore.Int()
+		if score > warnThreshold {
+			log.Warnf("Misbehaving peer %s: %s -- ban score is %d, "+
+				"it was not increased this time", p, reason, score)
+		}
+		return false
+	}
+
+	score := banScore.Increase(persistent, transient)
+	bm.banScores[p.Addr()] = banScore
```
Comment (davecgh) [inline connmgr/banmanager.go:190]: No need to set it again.  So long as `banScore` is a pointer (as my previous comment set it to) `Increase` updates it.

File: connmgr/banmanager.go [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+type banMgrPeer struct {
+	*peer.Peer
+
+	isWhitelisted bool
+}
+
+// BanManager represents a peer ban score tracking manager.
+type BanManager struct {
+	cfg       BanManagerConfig
+	peers     map[string]*banMgrPeer
```
Comment (davecgh) [inline connmgr/banmanager.go:47]: This should be keyed by `*peer.Peer` like `netsync.SyncManager.peers`.

Then, as also done in `netsync`, you could have a nice helper method for looking up the peer along with a warning if needed like so:

```Go
// lookupPeer returns the ban manager peer that maintains additional state for
// a given base peer.  In the event the mapping does not exist, a warning is
// logged and nil is returned.
//
// This function MUST be called with the ban manager mutex locked (for reads).
func (bm *BanManager) lookupPeer(p *peer.Peer) *banMgrPeer {
	bmp, ok := peers[p]
	if !ok {
		log.Warnf("Attempt to lookup unknown peer %s\nStack: %v", p,
			string(debug.Stack()))
		return nil
	}
	return bmp
}
```

File: connmgr/banmanager.go [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+
+// AddPeer adds the provided peer to the ban manager.
+func (bm *BanManager) AddPeer(p *peer.Peer) {
+	bmp := &banMgrPeer{
+		Peer:          p,
+		isWhitelisted: bm.IsPeerWhitelisted(p),
+	}
+
+	bm.mtx.Lock()
+	bm.peers[p.Addr()] = bmp
```
Comment (davecgh) [inline connmgr/banmanager.go:92]: This would then just become:

```Go
bm.peers[p]= bmp
```

File: connmgr/banmanager.go [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+	bm.banScores[p.Addr()] = &DynamicBanScore{}
+	bm.mtx.Unlock()
+}
+
+// LookupPeer returns the associated peer of the provided address and its
+// ban score.
+func (bm *BanManager) LookupPeer(addr string) (*peer.Peer, uint32) {
```
Comment (davecgh) [inline connmgr/banmanager.go:99]: Then this could just take the peer instance, be made to return the score, and renamed:

```Go
func (bm *BanManager) BanScore(p *peer.Peer) uint32 {
	bm.mtx.Lock()
	bmp := bm.lookupPeer(p)
	if bmp == nil {
		bm.mtx.Unlock()
	}
	score := bmp.banScore.Int()
	bm.mtx.Unlock()
	return score
}
```

File: connmgr/banmanager.go [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+// provided. Further, if the score is above the ban threshold, the peer will
+// be banned.
+func (bm *BanManager) AddBanScore(p *peer.Peer, persistent, transient uint32, reason string) bool {
+	// No warning is logged and no score is calculated if banning is disabled.
+	if bm.cfg.DisableBanning {
+		return false
+	}
+
+	bm.mtx.Lock()
+	bmp, ok := bm.peers[p.Addr()]
```
Comment (davecgh) [inline connmgr/banmanager.go:158]: `bmp := bm.lookupPeer(p)` etc

**Review by davecgh — CHANGES_REQUESTED** [review]
The updates look good overall and address almost everything.  The only major thing I see is that it really should be using the peer instances as the key.  I have several comments inline to explain better.  I didn't comment on every case, but it should be enough to show what I mean.

File: connmgr/banmanager.go [hunk 17]
```diff
@@ -81,39 +97,35 @@ func (bm *BanManager) IsPeerWhitelisted(p *peer.Peer) bool {
...
+	return bmp.isWhitelisted
+}
+
 // AddPeer adds the provided peer to the ban manager.
 func (bm *BanManager) AddPeer(p *peer.Peer) {
 	bmp := &banMgrPeer{
 		Peer:          p,
-		isWhitelisted: bm.IsPeerWhitelisted(p),
+		isWhitelisted: bm.isPeerWhitelisted(p, bm.cfg.WhiteList),
+		banScore:      DynamicBanScore{},
```
Comment (davecgh) [inline connmgr/banmanager.go:117]: This isn't needed since the zero value is immediately usable.

File: connmgr/banmanager.go [hunk 18]
```diff
@@ -125,10 +137,9 @@ func (bm *BanManager) BanPeer(p *peer.Peer) {
 	}
 
 	bm.mtx.Lock()
-	bmp, ok := bm.peers[p.Addr()]
-	if !ok {
+	bmp := bm.lookupPeer(p)
+	if bmp == nil {
```
Comment (davecgh) [inline connmgr/banmanager.go:141]: Since nothing else is using the lock, you can just unlock it before this if statement instead of the two places below.

File: internal/rpcserver/interface.go [hunk 19]
```diff
@@ -809,9 +812,8 @@ type BanManager interface {
 	// IsPeerWhitelisted checks if the provided peer is whitelisted.
 	IsPeerWhitelisted(p *peer.Peer) bool
 
-	// LookupPeer returns the associated peer of the provided address and
-	// its ban score.
-	LookupPeer(addr string) (*peer.Peer, uint32)
+	// BanScore returns the ban score of the provided peer.
+	BanScore(p *peer.Peer) uint32
```
Comment (davecgh) [inline internal/rpcserver/interface.go:816]: This is not needed.

**Review by davecgh — CHANGES_REQUESTED** [review]
I still need to test it and assert the new banning conditions, but otherwise, just a few nits left.  It's looking solid in terms of implementation.

File: internal/rpcserver/interface.go [hunk 20]
```diff
@@ -53,13 +53,12 @@ type Peer interface {
 	// LastPingNonce returns the last ping nonce of the remote peer.
 	LastPingNonce() uint64
 
+	// BanScore returns the ban score of the peer.
+	BanScore() uint32
```
Comment (davecgh) [inline internal/rpcserver/interface.go:57]: Also there was no reason to move or modify the comment on this one.

File: internal/rpcserver/rpcserver.go [hunk 21]
```diff
@@ -6541,6 +6541,9 @@ type Config struct {
 
 	// FiltererV2 defines the V2 filterer for the RPC server to use.
 	FiltererV2 FiltererV2
+
+	// BanMgr defines the peer manager for the RPC server to use.
+	BanMgr BanManager
```
Comment (davecgh) [inline internal/rpcserver/rpcserver.go:6546]: Actually, I don't think the `rpcserver` needs this of the associated interface at all since it's not used.

File: server.go [hunk 22]
```diff
@@ -1103,7 +1107,13 @@ func (sp *serverPeer) OnInv(p *peer.Peer, msg *wire.MsgInv) {
 
 // OnHeaders is invoked when a peer receives a headers wire message.  The
 // message is passed down to the net sync manager.
-func (sp *serverPeer) OnHeaders(_ *peer.Peer, msg *wire.MsgHeaders) {
+func (sp *serverPeer) OnHeaders(p *peer.Peer, msg *wire.MsgHeaders) {
+	// Ban peers sending empty headers requests.
+	if len(msg.Headers) == 0 {
```
Comment (davecgh) [inline server.go:1112]: Let's revert this change for this PR.  I'm fairly certain it is possible for an empty headers message to be sent in valid circumstances when reaching the chain tip during sync and this would end up banning them due to no fault of their own.

**Review by davecgh — CHANGES_REQUESTED** [review]
I haven't tested this yet, but reading the code, I don't see anywhere that it is adding the peers to the ban manager when they connect nor removing them when they disconnect.  I also don't see a function to actually ban the peer being specified in the config which means it would certainly panic for invoking a nil function.

**Comment by davecgh** [comment]
After going back and forth on this, please go ahead and put the code in `internal/staging/banmgr` for now and use it from there so we can having something to build on without putting it into `connmgr` until it's a bit more finalized because `connmgr` is an external-facing module and I don't want to cause more major module bumps unnecessarily, which is almost guaranteed until the ban manager gets fully fledged out and finalized over time a bit

**Review by davecgh — COMMENTED** [review]
Looks good with all the updates.  I'm going to test this out pretty heavily before approving.

**Comment by dnldd** [comment]
I'm going to update the test cases to cover the missing functionality you pointed out and implement accordingly.

**Review by davecgh — CHANGES_REQUESTED** [review]

File: server.go [hunk 23]
```diff
@@ -1100,12 +1075,6 @@ func (sp *serverPeer) OnBlock(p *peer.Peer, msg *wire.MsgBlock, buf []byte) {
 // accordingly.  We pass the message down to the net sync manager which will
 // call QueueMessage with any appropriate responses.
 func (sp *serverPeer) OnInv(p *peer.Peer, msg *wire.MsgInv) {
-	// Ban peers sending empty inventory requests.
```
Comment (davecgh) [inline server.go:1103]: This shouldn't be removed.

**Review by davecgh — CHANGES_REQUESTED** [review]

**Comment by davecgh** [comment]
Alright, so I'm still running into issues with this and banning still isn't even working properly.  Given the impending release, I really don't want to spend any more time on this.  Please revert all changes to everything except for the files under `internal/staging/banmanager` and squash everything down into a single commit, and update the PR and commit descriptions accordingly.

That way we'll have the primary work on the ban manager itself as a starting point for moving forward, but it will not be used for now until we've had time to iterate on the design and properly integrate it.

File: connmgr/go.mod [hunk 24]
```diff
@@ -6,3 +6,9 @@ require (
 	github.com/decred/dcrd/wire v1.4.0
 	github.com/decred/slog v1.1.0
 )
+
+replace (
```
[the remark this answers is not in the document]
Comment (dnldd) [reply connmgr/go.mod:10]: Reverting.

File: server.go (hunk 23 above)
Comment (dnldd) [reply server.go]: Reverted.

**Review by davecgh — APPROVED** [review]
